src/app/api/public/contracts/[contractName]/[usersContractVersion]/[functionName]/route.ts

```diff
@@ -93,92 +90,67 @@ async function postHandler(req: NextRequestWithAuth, { params: { contractName, u
 
     const userId = req.userId;
     const developerId = req.developerId;
-    const accountData = !!userId
-      ? await developersUserAccountModel.findUnique({ where: { id: userId } })
-      : await developerAccountModel.findUnique({ where: { id: developerId } });
-
+    const targetAccount = !!userId
+      ? await developersUserAccountModel.findUnique({ where: { id: userId }, select: {id: true} })
+      : await developerAccountModel.findUnique({ where: { id: developerId }, select: {id: true} });
+    const { account } = await getAccountInstance(!!userId ? { userId: targetAccount.id } : { developerId: targetAccount.id });
     if (targetFunction.type === "read") {
-      // 🏗️ TODO: call the contract as the user, not operator
-      const contract = new Contract(contractsInterfaces[contractName].abi, smartContract?.address);
+      const contract =  new Contract(contractsInterfaces[contractName].abi, smartContract?.address)
       let result = await contract[functionName](...Object.values(validBody));
 
-      // 🏗️ TODO: read the Cairo's type of the function's input, distinguish between Contract Address, and BigInt, for
-      // better display of result if (typeof result === "bigint") { console.log("🔮 result: ", (result.toString()))
-      // result = `0x${result.toString(16)}`; result = BigInt(result).toString(16); }
       if (typeof result === "bigint") {
         result = `0x${result.toString(16)}`;
       }
       return NextResponse.json(
-        { result: result, type: typeof result },
-        { status: StatusCodes.OK }
+          { result, type: typeof result },
+          { status: StatusCodes.OK }
       );
     } else {
-      const keys = await getVaultItem(accountData.vaultKey, "privateKey");
-      const { walletAddress, privateKey } = retrieveWalletCredentials(keys);
-      const account = new Account(provider, walletAddress, privateKey);
-      const calldata = getGaslessTransactionCallData({
-        method: functionName,
-        contractAddress: contract.address,
-        body,
-        abiFunctions: functions
-      });
-
-      const tx = await gaslessTransaction(account, calldata);
-      const txReceipt = !!tx?.transactionHash ? await provider.waitForTransaction(tx.transactionHash) as any : null;
+      const transactionResult = await gaslessTransactionWithFallback(account, functionName, contract, body, functions);
+      const txReceipt = !!transactionResult?.txHash
+        ? ((await provider.waitForTransaction(
+          transactionResult?.txHash,
+          )) as any)
+        : null;
+
+      if (!!transactionResult.error) {
+        return NextResponse.json(
+          {
```

Comment:
here you go with

```
        {
            result: transactionResult,
          },
```

while 10 lines above you go with 
          
          
```
 { result, type: typeof result },
```
pls adjust